gnome-initial-setup/pages/language/cc-language-chooser.c

```diff
@@ -145,10 +148,80 @@ language_widget_new (const char *locale_id,
         return widget->box;
 }
 
+typedef struct {
+        CcLanguageChooser *chooser;
+        GtkWidget         *row;
+} ScrollData;
+
+static inline gint
+get_selected_language_row_y (GtkWidget *row)
+{
+        GtkAllocation alloc;
+
+        gtk_widget_get_allocation (row, &alloc);
+
+        return alloc.y + alloc.height / 2.0;
+}
+
+static void
+update_scroll_position (ScrollData *data)
+{
+        CcLanguageChooser *self;
+        GtkAdjustment *vadjustment;
+        GtkWidget *row;
+        gdouble page_increment;
+        gdouble real_value;
+
+        self = data->chooser;
+        row = data->row;
+
+        vadjustment = gtk_scrolled_window_get_vadjustment (GTK_SCROLLED_WINDOW (self->priv->language_scrolledwindow));
+        page_increment = gtk_adjustment_get_page_increment (vadjustment);
+        real_value = get_selected_language_row_y (row) - page_increment / 2.0;
+
+        /*
+         * Since GtkScrolledWindow tries to follow the focused row of a listbox,
+         * make sure that the current row is focused.
+         */
+        gtk_widget_grab_focus (row);
+
+        gtk_adjustment_set_value (vadjustment, real_value);
+
+        self->priv->initial_scroll = TRUE;
+
+        if (self->priv->scroll_handler_id > 0) {
+                g_signal_handler_disconnect (self->priv->language_scrolledwindow,
+                                             self->priv->scroll_handler_id);
+                self->priv->scroll_handler_id = 0;
+        }
+}
+
+static void
+schedule_scroll (CcLanguageChooser *self,
+                 GtkWidget         *row)
+{
+        ScrollData *data;
+
+        data = g_new0 (ScrollData, 1);
+        data->row = row;
+        data->chooser = self;
```

Comment:
At the moment this data struct is leaked; but I think I would prefer it if you just saved the row to scroll to inside the private structure instead of using a separate one.